src/IO/H5/Helpers.cpp

```diff
@@ -19,30 +23,129 @@
+template<size_t Rank, typename T>
+struct VectorToArray;
+
+template <typename T>
+struct VectorToArray<0, T> {
+  static T apply(std::vector<T> raw_data,
+                 const std::array<hsize_t, 1>& /*size*/) noexcept {
+    return raw_data[0];
+  }
+};
+
+template <typename T>
+struct VectorToArray<1, T> {
+  static T apply(T raw_data, const std::array<hsize_t, 1>& /*size*/) noexcept {
+    return raw_data;
+  }
+};
+
+template <>
+struct VectorToArray<2, DataVector> {
+  static DataVector apply(DataVector raw_data,
+                          const std::array<hsize_t, 2>& /*size*/) noexcept {
+    return raw_data;
+  }
+};
+
+template <typename T>
+struct VectorToArray<2, boost::multi_array<T, 2>> {
+  DEBUG_STATIC_ASSERT(cpp17::is_fundamental_v<T>,
+                      "VectorToArray is optimized for fundamentals. Need to "
+                      "use move semantics for handling generic data types.");
+  static boost::multi_array<T, 2> apply(
+      const std::vector<T>& raw_data,
+      const std::array<hsize_t, 2>& size) noexcept {
+    boost::multi_array<T, 2> temp(boost::extents[size[0]][size[1]]);
+    for (size_t i = 0; i < size[0]; ++i) {
+      for (size_t j = 0; j < size[1]; ++j) {
+        temp[i][j] = raw_data[j + i * size[1]];
+      }
+    }
+    return temp;
+  }
+};
+
+template <>
+struct VectorToArray<3, DataVector> {
+  static DataVector apply(DataVector raw_data,
+                          const std::array<hsize_t, 3>& /*size*/) noexcept {
+    return raw_data;
+  }
+};
+
+template <typename T>
+struct VectorToArray<3, boost::multi_array<T, 3>> {
+  DEBUG_STATIC_ASSERT(cpp17::is_fundamental_v<T>,
+                      "VectorToArray is optimized for fundamentals. Need to "
+                      "use move semantics for handling generic data types.");
+  static boost::multi_array<T, 3> apply(
+      const std::vector<T>& raw_data,
+      const std::array<hsize_t, 3>& size) noexcept {
+    boost::multi_array<T, 3> temp(boost::extents[size[0]][size[1]][size[2]]);
+    for (size_t i = 0; i < size[0]; ++i) {
+      for (size_t j = 0; j < size[1]; ++j) {
+        for (size_t k = 0; k < size[2]; ++k) {
+          temp[i][j][k] = raw_data[k + j * size[2] + i * size[2] * size[1]];
+        }
+      }
+    }
+    return temp;
+  }
+};
+}  // namespace
+
+namespace {
+// Implementation of 'write_data'. Common to both DataVector and std::vector
+template <typename T>
+void write_data_impl(const hid_t group_id, T data,
+                     const std::vector<size_t>& extents,
+                     const std::string& name) noexcept {
   const std::vector<hsize_t> dims(extents.begin(), extents.end());
   const hid_t space_id = H5Screate_simple(dims.size(), dims.data(), nullptr);
   CHECK_H5(space_id, "Failed to create dataspace");
-  const hid_t contained_type = h5::h5_type<std::decay_t<decltype(data[0])>>();
+  const hid_t contained_type =
+      h5::h5_type<typename tt::get_fundamental_type_t<T>>();
```

Comment:
You don't need `typename` if you are not accessing a nested type. Please do this everywhere else too